content/blog/2019-10-19-loop-invariant-code-motion/index.md

```diff
@@ -0,0 +1,297 @@
+
+Once these instructions are identified, we check if these instructions can be hoisted.
+We move hoisted instructions into a preheader basic block and rename jump targets accordingly.
+After post-processing, we have our new instruction list!
+
+
+<img src="plan.jpeg" style="width: 100%">
+
+
+Skip to the end to optimize your very own `bril` program!
+
+# Loop
+
+All loops considered here are **natural loops**—that is, a cycle with one entry and a **back-edge**. Back-edges are defined as an edge $A \longrightarrow B$ for tail $A$ and head $B$, such that $B$ dominates $A$.  Natural loops are then defined as the smallest set of vertices $L$ with $A,B \in L$ such that for each vertex $v \in L$ we have $v=B$ or PREDS($v$)$\subseteq L$.
+
+In essence, a back-edge is what brings us from the tail of the loop $A$ to the beginning $B$.
+The cycle surrounding this backedge is our loop, where the entrypoint is the start of the cycle.
+If the loop has only one entry, it is a natural loop. Below is an illustration of a natural loop and its back-edge. The natural loop is highlighted, with a labeled back-edge.
+
+<img src="natloop.jpeg" style="width: 70%">
+
+In the case that an edge connected *entry* and *body*, this would no longer be a natural loop.
+ 
+
+### Detecting Natural Loops
+
+
+To find the loop invariant code, first we must detect all natural loops. To accomplish this, we make use of control flow graphs from `cfg.py` and dominator trees from `dom.py` within our three functions. Back-edges are identified with `get_backedges`, and `loopsy` finds the natural loop associated with an input back-edge.
+
+
+```python
+### detect back-edges
+def get_backedges(successors,domtree):
+    backedges = set()
+    for source,sinks in successors.items():
+        for sink in sinks:
+            if sink in domtree[source]:
+                backedges.add((source,sink))
+
+    return backedges
+
+
+### get natural loops
+def loopsy(source,sink,predecessors):
+    worklist = [source]
+    loop = set()
+    while len(worklist)>0:
+        current = worklist.pop()
+        pr = predecessors[current]
+        for p in pr:
+            if not(p in loop or p==sink):
+                loop.add(p)
+                worklist.append(p)
+    loop.add(sink)
+    loop.add(source)
+
+    return loop
+```
+
+
+
+### Detecting loop invariants
+An instruction within a natural loop is marked loop-invariant if its arguments are defined outside of the natural loop. Alternately, if the instruction’s arguments are defined once—and that definition is loop invariant—then the instruction may be marked as loop-invariant. Our goal is to find these loop invariants so that we may mark them as movable. Iteratively checking these two conditions, we converge on a list of invariant instructions.
```

Comment:
You've used "loop invariant" without the hyphen elsewhere—maybe stick to that? Lots of style guides prefer hyphens only before the verb, so "marked as loop invariant" is a great place to _not_ use a hyphen.